// approuter/lib/passport/jwt-refresh.js
'use strict';

var request = require('request');
var VError = require('verror').VError;

var loggerUtils = require('../utils/logger');
var passportUtils = require('./utils');

var tracer = loggerUtils.getTracer();
var MAX_ATTEMPTS_NUMBER = 3;

module.exports = {
  executeAfter: function(fn, timeout) {
    setTimeout(fn, timeout);
  },

  refreshToken: function(app, session) {
    tracer.path('Entering refreshToken');
    var jwtRefresh = app.get('routerConfig').jwtRefresh;
    var tokenRefreshTimestamp = session.oauthToken.expiryDate - toMilliseconds(jwtRefresh);

    var msBeforeRefresh = tokenRefreshTimestamp - Date.now();
    if (msBeforeRefresh < 0) {
      return tracer.error('jwtRefresh is greater than the the token validity period!');
    }

    var memoryStore = app.get('memoryStore');
    memoryStore.getSessionTimeout(session.id, function(err, sessionTimeout) {
      if (err || Date.now() + toMilliseconds(sessionTimeout) < tokenRefreshTimestamp) {
        return err && tracer.debug(err);
      }

      memoryStore.update(session.id, function(newSession) {
        if (!newSession || newSession.jwtRefreshStarted) {
          return;
        }
        tracer.info('Updating jwtRefreshStarted to true.');
        newSession.jwtRefreshStarted = true;
        tracer.info('Refreshing jwt in', msBeforeRefresh, 'ms.');
        module.exports.executeAfter(function() {
          doRefresh(app, newSession);
        }, msBeforeRefresh);
      }, false);
    });
  }
};

function toMilliseconds(minutes) {
  return minutes * 60 * 1000;
}

function doRefresh(app, session) {
  var routerConfig = app.get('routerConfig');
  var uaaOptions = routerConfig.uaaOptions;
  var requestOptions = {
    url: uaaOptions.tokenURL,
    form: {
      'grant_type': 'refresh_token',
      'refresh_token': session.oauthToken.refreshToken
    },
    auth: {
      user: uaaOptions.clientID,
      pass: uaaOptions.clientSecret
    }
  };

  doRefreshAttempt(requestOptions, app, session, 1);
}

function doRefreshAttempt(requestOptions, app, session, attemptNumber) {
  tracer.path('Entering doRefreshAttempt');
  var timeout = session.oauthToken.expiryDate - Date.now();
  if (timeout < 0) {
    return tracer.info('JWT could not be refreshed!');
  }
  requestOptions.timeout = timeout;
  callUaa(requestOptions, attemptNumber, function uaaResponse(err, uaaResponse) {
    if (err) {
      tracer.warning(err);
      if (attemptNumber === MAX_ATTEMPTS_NUMBER) {
        return tracer.error('Unable to refresh the JWT, number of attempts done: %d', attemptNumber);
      }
      return doRefreshAttempt(requestOptions, app, session, attemptNumber + 1);
    }
    var updatedSession;
    app.get('memoryStore').update(session.id, function updateSession(newSession) {
      if (!newSession) {
        return;
      }
      var options = {
        accessToken: uaaResponse.access_token,
        expiresIn: uaaResponse.expires_in,
        refreshToken: uaaResponse.refresh_token,
        scope: uaaResponse.scope
      };
      passportUtils.setTokenToSession(newSession, options);
      newSession.jwtRefreshStarted = false;
      updatedSession = newSession;
      tracer.info('Token refreshed!');
    }, false);
    updatedSession && module.exports.refreshToken(app, updatedSession); // schedule next refresh if needed
  });
}

function callUaa(requestOptions, attemptNumber, cb) {
  request.post(requestOptions, function onResponse(err, res, body) {
    if (err || res.statusCode !== 200) {
      var msg = 'Attempt ' + attemptNumber + ' of jwt refresh failed! ' + (err ? err : 'Status code: ' + res.statusCode) + ' responseBody: ' + body;
      return cb(new Error(msg));
    }
    try {
      var responseJSON = JSON.parse(body);
    } catch (err) {
      return cb(new VError(err, 'Bad response from UAA when refreshing JWT - not a JSON'));
    }
    if (!responseJSON.access_token || !responseJSON.expires_in || !responseJSON.refresh_token) {
      return cb(new Error('Bad response from UAA when refreshing JWT - not all JWT fields are present'));
    }

    cb(null, responseJSON);
  });
}
